Sources/MessagingInApp/Gist/Managers/MessageManager.swift

```diff
@@ -39,35 +108,25 @@ class MessageManager: EngineWebDelegate {
         }
     }
 
-    func showMessage(position: MessagePosition) {
-        elapsedTimer.start(title: "Displaying modal for message: \(currentMessage.messageId)")
-        messagePosition = position
+    // MARK: Timer determining how long message took to load.
+
+    // The manager subclasses are expected to call these functions to determine how long the messages took to load.
+
+    func startLoadingMessage() {
+        elapsedTimer.start(title: "Loading message with id: \(currentMessage.messageId)")
+    }
+
+    func doneLoadingMessage() {
+        elapsedTimer.end()
     }
 
     func getMessageView() -> GistView {
         isMessageEmbed = true
         return gistView
     }
 
-    private func loadModalMessage() {
-        if messageLoaded {
-            modalViewManager = ModalViewManager(gistView: gistView, position: messagePosition)
-            modalViewManager?.showModalView { [weak self] in
-                guard let self = self else { return }
-                self.delegate?.messageShown(message: self.currentMessage)
-                self.elapsedTimer.end()
-            }
-        }
-    }
-
     func dismissMessage(completionHandler: (() -> Void)? = nil) {
-        if let modalViewManager = modalViewManager {
-            modalViewManager.dismissModalView { [weak self] in
-                guard let self = self else { return }
-                self.delegate?.messageDismissed(message: self.currentMessage)
-                completionHandler?()
-            }
-        }
+        // expect subclass implements this.
```

Comment:
I am honestly not big fan of subclassing when there are different approaches that makes the code easier to follow.
MessageManager can simply be a protocol with two different implementations if needed. The actual implementations can use similar utility functions for the common logic.
Inheritance is a very strong relation and unless absolutely a must, we can almost always find simpler approaches

Reply:
See [my latest update](https://github.com/customerio/customerio-ios/pull/717#issuecomment-2135841512).